ctapipe/image/muon/processor.py

```diff
@@ -0,0 +1,257 @@
+
+    quality_criteria = List(
+        Tuple(Unicode(), Unicode()),
+        default_value=[
+            ("radius_not_nan", "np.isfinite(ring.radius.value)"),
+            ("min_pixels", "np.count_nonzero(mask) > 50"),
+            ("ring_containment", "parameters.containment > 0.5"),
+        ],
+        help=QualityQuery.quality_criteria.help,
+    ).tag(config=True)
+
+
+class MuonProcessor(TelescopeComponent):
+    """
+    Takes cleaned images and extracts muon rings. Should be run after ImageProcessor.
+    """
+
+    completeness_threshold = FloatTelescopeParameter(
+        default_value=30.0, help="Threshold for calculating the ``ring_completeness``"
+    ).tag(config=True)
+
+    ratio_width = FloatTelescopeParameter(
+        default_value=1.5,
+        help="Ring width for intensity ratio computation in units of pixel diameter",
+    ).tag(config=True)
+
+    pedestal = FloatTelescopeParameter(
+        help="Pedestal noise rms", default_value=1.1
+    ).tag(config=True)
+
+    def __init__(self, subarray, **kwargs):
+        super().__init__(subarray, **kwargs)
+        self.dl1_query = ImageParameterQuery(parent=self)
+        self.ring_query = RingQuery(parent=self)
+
+        frame = TelescopeFrame()
+        self.geometries = {
+            tel_id: tel.camera.geometry.transform_to(frame)
+            for tel_id, tel in self.subarray.tel.items()
+        }
+        self.fov_radius = {
+            tel_id: geometry.guess_radius()
+            for tel_id, geometry in self.geometries.items()
+        }
+
+        self.ring_fitter = MuonRingFitter(parent=self)
+
+        self.intensity_fitter = MuonIntensityFitter(subarray=subarray, parent=self)
+
+    def __call__(self, event: ArrayEventContainer):
+        for tel_id in event.dl1.tel:
+            self._process_telescope_event(event, tel_id)
+
+    def _process_telescope_event(self, event, tel_id):
+        """
+        Extract and process a ring from a single image.
+
+        Parameters
+        ----------
+        event: ArrayEventContainer
+            Collection of all event information
+        tel_id: int
+            Telescope ID of the instrument that has measured the image
+        """
+        event_index = event.index
+        event_id = event_index.event_id
+
+        if self.subarray.tel[tel_id].optics.n_mirrors != 1:
+            self.log.warning(
+                f"Skipping non-single mirror telescope {tel_id},"
+                " muon-fitting for dual-mirror telescope is currently"
+                " not supported. Exclude dual mirror telescopes via setting"
+                " 'EventSource.allowed_tels'."
+            )
+            event.muon.tel[tel_id] = INVALID
+            return
+
+        self.log.debug(f"Processing event {event_id}, telescope {tel_id}")
+        dl1 = event.dl1.tel[tel_id]
+        image = dl1.image
+        mask = dl1.image_mask
+        if mask is None:
+            mask = image > 0
+
+        checks = self.dl1_query(p=dl1.parameters)
```

Comment:
Can we call this `parameters=...`, I don't like the `p`?

Reply:
The ring-query also has a `parameters` attribute, this might lead to confusion. Maybe `dl1_parameters` or `dl1_params` for shorter notation?